### ISSUE TITLE
Memory is slowly leaking

### ISSUE THREAD

**Issue by dmpetrov** [python-pillow/Pillow#2019]
Memory is slowly leaking when process many image files one by one.

Leak size:
100K images --> ~100MB
300K images --> ~200MB
1M images --> ~1GB

Here you can find reproducing code examples: https://github.com/PyWavelets/pywt/issues/180

**Comment by hugovk** [comment]
@dmpetrov Please can you give a minimal code example reproducing the problem that uses Pillow but no third-party libraries like numpy or pywt? Thanks!

**Comment by dmpetrov** [comment]
```
import os
import PIL
from PIL import Image

dir = "/Volumes/Seagate/storage/kaggle/avito/images"
onlyfiles = []
dirs = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
for d in dirs:
    filedir = os.path.join(dir, str(d))
    files = [f for f in os.listdir(filedir) if os.path.isfile(os.path.join(filedir, f))]
    filepaths = map( lambda x: os.path.join(filedir, x), files )
    onlyfiles += list(filepaths)

s = 0
for i in range(len(onlyfiles)):
    fname = onlyfiles[i]
    image = PIL.Image.open(fname)
    # Do something
    s += image.size[0]*image.size[1]

    if i % 10000 == 0:
        print('iteration ', i)

input("Press Enter to continue...")
```

Input data: image dataset (10M small images) from Kaggle competition https://www.kaggle.com/c/avito-duplicate-ads-detection

Result: It starts with 221MB for the file names.
500 images - 374MB
1M images - 510MB

gc() inside the loop did not help.

**Comment by damaestro** [comment]
Do we know what release this started in? I'd like to pin to a lower release to mitigate any production impact without having to set memory scaling alarm/actions.

**Comment by homm** [comment]
@dmpetrov I can't reproduce this on Ubuntu 14.04 with Python 2.7 on Pillow master. Could you tell your Pillow, Python, OS, libjpeg versions?

@damaestro This is not confirmed yet.

**Comment by homm** [comment]
@dmpetrov I'v tried Python 2.7 and 3.4 under OS X. I can confirm the leak in 3.4.

**Comment by homm** [comment]
So, this is a minimal test case, which I came to:

https://gist.github.com/homm/c6a79ce7b445f47a74a4d296742a5af8

As you can there is no Pillow at all! I'm pretty sure this is pure Python 3 memory leak.

The script accepts path to the root folder with tons of files as argument and iterates all files in the folder twice: First time it opens the files and calls `.close()` method. At this stage, memory usage doesn't grow. Second time the script doesn't call `.close()` method and memory usage constanlty increasing. I've tested Python 3.4 and 3.5 on OS X and Python 3.4 on Ubuntu 14.04. It doesn't appear on Python 2.7.

This is **Python's** memory leak because there are no pointers to the file objects from the application space. This is most likely **only memory** leak. At least `lsof` doesn't show any opened files after script's end. This leak appears only on different files. It doesn't work if we are reopening one file again and again. Average memory consumption is 370 bytes per file.

As I understand we can't fix it on Pillow's level because we are assuming that file pointer may be shared with the other code. There are two possibilities: It can be fixed in Python itself. Or it can be fixed on the application level. Something like this:

```python
f = open(filename, 'rb')
image = Image.open(f)
image.load()
f.close()
```

instead of `Image.open(filename)`

**Comment by homm** [comment]
I want to Invite @asvetlov to the thread. Maybe he could clarify something.

**Comment by hugovk** [comment]
What happens when using `with open(file.path, 'rb', 0) as f:`?

**Comment by dmpetrov** [comment]
@homm thank you for the investigation. It is becoming more and more interesting

Yes, I use python 3.5. As far as I know, Python does not guarantee that file will be closed. So, it might be Python 3+ "feature", not a bag.

It would be great to have opinions of Python experts.

**Comment by homm** [comment]
> Python does not guarantee that file will be closed.

Of course [it guarantees](https://docs.python.org/3/library/io.html#io.IOBase.__del__). Moreover, it guarantees that exactly `.close()` method will be called. It doesn't guarantee when this will be done, that is all.

But as I said we are not speaking about the file descriptors leak. All files are closed. The problem is there is a memory leak.

**Comment by damaestro** [comment]
I'm seeing the issue with Py 2.7 running thumbor, so I don't think this is isolated to Py 3.x.

![screenshot from 2016-07-14 15-22-59](https://cloud.githubusercontent.com/assets/5994916/16856341/df50ff9a-49d6-11e6-9999-fc5856466209.png)

**Comment by homm** [comment]
@damaestro

> so I don't think **this** is isolated to Py 3.x.

Any reasons why you are thinking this is **the same** leak? Memory leaks are not isolated to Python 3 or Python or any other platform or library. In this thread, we are discussing leak in Python 3 which affects Pillow users. You can report thumbor leaks in [appropriate place](https://github.com/thumbor/thumbor/issues).

**Comment by homm** [comment]
@hugovk `with file` statement correctly frees all memory like `.close()` do.

**Comment by asvetlov** [comment]
Guys, sorry.
I really don't know what the difference between `f.close()`/`with f` and `del f` except deleting of non-closed file should raise a warning along with actual file closing.

**Comment by asvetlov** [comment]
But explicit resource cleanup (either via `with` statement or by `.close()` call is considered as good practice).

**Comment by homm** [comment]
> deleting of non-closed file should raise a warning

I haven't met this in the docs. Could you elaborate?

I'll try to simplify the test and fill report to the Python bug tracker at holidays.

Also, I'm sure that we can at least explicit close files which were opened in `Image.open` by the filename. This should fix bug on our side for most of the users.

**Comment by homm** [comment]
Looks like this is exactly warnings issue. If I'm adding `f._dealloc_warn(f)` right before close, the memory is also leakings.

The memory is allocated to store all already occurred warnings to prevent occurrence second time.  Horrible decision in my opinion. Not closed file files **might** lead to memory leaks, while storing error message for all files is absolutely guaranteed memory leak.

**Comment by damaestro** [comment]
@homm no reason. I'll continue to follow this discussion and if I find something indicating it's an issue with Pillow I'll speak up. Thanks.

**Comment by homm** [comment]
The issue in the Python tracker: http://bugs.python.org/issue27535

**Comment by homm** [comment]
Sorry, moving to 3.5.0 cause of lack of time. The workaround for this is using file objects instead of file names:

```python
with open(filename, 'rb') as f:
    image = Image.open(f)
    # We need to consume the whole file inside the `with` statement
    image.load()
    . . .
# Unref here to be sure that there is nothing happens
# with the image after file is closed
image = None
```

**Comment by okor** [comment]
@homm Sorry, it's a little unclear what the status of this bug is. In your last comment you said you were moving this issue to version 3.5.0 due to lack of time ... but I don't see a release past 3.4.2. 

And it seems like there is a duplicate bug, for version 4 over here https://github.com/python-pillow/Pillow/issues/2361

Can you confirm that the bug is still present? Thank you

**Comment by homm** [comment]
@wiredfool Does it fixed?

**Comment by homm** [comment]
It was partly fixed [in 4.1.0](http://pillow.readthedocs.io/en/latest/releasenotes/4.1.0.html#closing-files-when-opening-images), but for some formats (GIF, for example), [the problem is still exists](https://github.com/python-pillow/Pillow/issues/2716).

**Comment by radarhere** [comment]
Is this now fixed on master?

**Comment by radarhere** [comment]
If anyone in this issue has thoughts on #3577, that could be helpful.

**Comment by hugovk** [comment]
> The issue in the Python tracker: http://bugs.python.org/issue27535

This has been fixed in Python 3.7.

**Comment by radarhere** [comment]
Note that implicitly closing an image's file has now been deprecated - https://pillow.readthedocs.io/en/stable/releasenotes/6.1.0.html#deprecations

**Comment by radarhere** [comment]
Support has now been removed for implicitly closing an image's file - #3577

**Comment by SaschaHeyer** [comment]
It's hard to follow this issue can someone explain what is the suggested solution?

**Comment by radarhere** [comment]
@SaschaHeyer I think the problem discussed here is likely resolved. If you have a situation, I'd recommend that you open a new issue with a self-contained example

**Comment by SaschaHeyer** [comment]
@radarhere 
Thank you for your quick response.

I wanted to know how the issue is resolved. 
What steps are required to solve?

**Comment by hugovk** [comment]
Let's close this, and we can re-open if needed, or open a new issue.

**Comment by radarhere** [comment]
> @radarhere
> Thank you for your quick response.
> 
> I wanted to know how the issue is resolved.
> What steps are required to solve?

To clarify, I meant that I suspect it is resolved in the current version of Pillow, fixed by changes to our code, not to yours.

**Comment by Mvbbb** [comment]
Is this problem got fixed in python2.7 pillow6.2.2

**Comment by radarhere** [comment]
It's not completely clear in this issue what the problem was, or if/how it was fixed. Here are some notes though
- The memory leak in Python itself [wasn't able to be reproduced in Python 2.7](https://github.com/python-pillow/Pillow/issues/2019#issuecomment-232556029).
- #3261 may have helped, and that was part of Pillow 5.4.0.
- https://github.com/python-pillow/Pillow/pull/3577 only removed support for implicitly closing an image's file in Pillow 7.0.0, but you could still call `close()` on images or use a context manager previously.

**Comment by Mvbbb** [comment]
> It's not completely clear in this issue what the problem was, or if/how it was fixed. Here are some notes though在这个问题上，并不完全清楚问题是什么，或者是否/如何解决。不过，这里有一些注意事项
> 
> * The memory leak in Python itself [wasn't able to be reproduced in Python 2.7](https://github.com/python-pillow/Pillow/issues/2019#issuecomment-232556029).Python 本身的内存泄漏无法在 Python 2.7 中重现。
> * [Resolve __fp when closing and deleting #3261](https://github.com/python-pillow/Pillow/pull/3261) may have helped, and that was part of Pillow 5.4.0. [Resolve __fp when closing and deleting #3261](https://github.com/python-pillow/Pillow/pull/3261) 可能有所帮助，这是 Pillow 5.4.0 的一部分。
> * [Improve handling of file resources #3577](https://github.com/python-pillow/Pillow/pull/3577) only removed support for implicitly closing an image's file in Pillow 7.0.0, but you could still call `close()` on images or use a context manager previously. [Improve handling of file resources #3577](https://github.com/python-pillow/Pillow/pull/3577) 仅在 Pillow 7.0.0 中删除了对隐式关闭映像文件的支持，但 `close()` 您仍然可以调用映像或使用上下文管理器。

Our service encountered some memory leaks and continuous memory growth when using Pillow to extract frames from webp animations. Can you identify any potential issues with the following code? Deeply grateful!
- python: 2.7.9
- pillow: 6.2.2

``` python
def extract_gif_use_pillow(image):
    frames = []
    try:
        with Image.open(StringIO(image)) as im:
            with BytesIO() as frame:
                while True:
                    try:
                        frame.seek(0)
                        im.save(frame, format='WEBP', quality=100)
                        frames.append(frame.getvalue())
                        # move to next frame
                        im.seek(im.tell() + 1)
                    except EOFError:
                        break
    except Exception as e:
        logger.error('extract_gif_use_pillow got exception: %s', e)
    gc.collect()
    logger.info('extract_gif_use_pillow got frames num: %s', len(frames))
    return frames
```

**Comment by radarhere** [comment]
I see no obvious problems, but I do recommend that you use a version of Python that is still maintained, and a more recent version of Pillow. Pillow has also made significant improvements to how it reads GIF images.

You may like to have a read of https://github.com/python-pillow/Pillow/issues/7935#issuecomment-2031804237. It is possible that is the explanation for your memory issues.
